src/services/completions.ts

```diff
@@ -1157,7 +1157,10 @@ namespace ts.Completions {
                     // We don't support overloads in object literals.
                     return undefined;
                 }
-                let typeNode: FunctionTypeNode | TypeNode | undefined = checker.typeToTypeNode(type, enclosingDeclaration, builderFlags, codefix.getNoopSymbolTrackerWithResolver({ program, host }));
+                const effectiveType = type.flags & TypeFlags.Union && (type as UnionType).types.length < 10
+                    ? checker.getUnionType((type as UnionType).types, UnionReduction.Subtype)
+                    : type;
+                let typeNode = checker.typeToTypeNode(effectiveType, enclosingDeclaration, builderFlags, codefix.getNoopSymbolTrackerWithResolver({ program, host }));

```

Comment:
Did you consider `ts.codefix.createSignatureDeclarationFromSignature` instead? It handles `tryGetAutoImportableReferenceFromTypeNode` for you, and seems like it should handle this union reduction, though I can’t quite tell for sure from skimming the code.

Reply:
The `typeToTypeNode` route seems odd to me because it could potentially do a lot of work to serialize something that doesn’t end up being a function and then we throw it away. You’re already checking that there’s not _more_ than 1 signature; isn’t checking that there’s _exactly_ 1 what you want? Or is that subtly different in a way I’m not thinking of?

Reply:
I found that using the signature instead of the type is tricky. In the cases where the symbol has a union type, like `M: (x: number) => void | (y: string) => void`, we get a single signature that is `M(x: never): void`, so it's not clear to me how to do things based on the signature.
I still also plan on adding a check that if the type is a union, it should only have one function type component.

Reply:
Would it at least make sense to ensure there’s 1 signature above, instead of 0 or 1?